Modules/_decimal/_decimal.c

```diff
@@ -3311,6 +3326,19 @@ dec_format(PyObject *dec, PyObject *args)
         }
     }
 
+    if (no_neg_0 && mpd_isnegative(MPD(dec)) && !mpd_isspecial(MPD(dec))) {
+        /* round into a temporary and clear sign if result is zero */
+        mpd_uint_t dt[MPD_MINALLOC_MAX];
+        mpd_t tmp = {MPD_STATIC|MPD_STATIC_DATA,0,0,0,MPD_MINALLOC_MAX,dt};
+        mpd_qrescale(&tmp, MPD(dec), -spec.prec, CTX(context), &status);
```

Comment:
This doesn't seem right. We want to replicate the rounding of `mpd_qformat_spec`, but depending on the presentation format, we might want to round to a number of significant figures instead of to a fixed exponent like `qrescale` does. For example, I get the following behaviour on this PR branch:

```python
>>> from decimal import Decimal
>>> format(Decimal("-0.001"), ".2e")
'-1.00e-3'
>>> format(Decimal("-0.001"), "z.2e")  # Expected '-1.00e-3'
'1.00e-3'
```

`%`-formatting is also mishandled:

```python
>>> format(Decimal("-0.001"), ".2%")
'-0.10%'
>>> format(Decimal("-0.001"), "z.2%")  # Expected '-0.10%'
'0.10%'
```

Also, I think `mpd_set_positive` isn't going to result in the right thing if we have a directed rounding mode:

```python
>>> from decimal import Decimal, getcontext, ROUND_CEILING
>>> getcontext().rounding = ROUND_CEILING
>>> format(Decimal("-0.001"), ".2f")
'-0.00'
>>> format(Decimal("-0.001"), "z.2f")  # Expected `0.00`
'0.01'
```

In general, the match-the-rounding approach feels too fragile to me. Instead, can we do post-formatting manipulation of the result? I know that detecting zero values post-formatting is not going to be trivial. It looks as though the biggest pain will be dealing with the fill character. We could update the logic so that we use a dummy fill character that won't interfere with the post-formatting-zero-detection, and then always replace that fill character post-formatting.

On the plus side, this problem reinforces the value of PEP 682 by demonstrating that it's non-trivial to independently match the rounding performed by the formatting.

Reply:
>  I know that detecting zero values post-formatting is not going to be trivial.

Detection doesn't seem that bad-- it may just be the fill substitution and checking for the sequence `r'-0[0.]*($|[^1-9])'`.

But knowing how to remove or substitute the minus after it's detected seems hard, and relies on width, alignment, and fill-- I wouldn't want to take that on.

I didn't follow why the directed rounding mode affects `mpd_set_positive()`.  There must be some way to remove the sign and reformat... would multiplying by -1 work?

Reply:
> There must be some way to remove the sign and reformat...

(I really wanted to avoid having to run though formatting twice, but I guess it would be limited to the relatively rare case where `z` is enabled and the 1st format pass is actually negative zero.  And this will go away if/when the mpd formatting code is pulled into `_decimal.c`.)

Reply:
At this point, I'm desperately seeking ways to avoid forking mpd formatting for this feature.

Ideas (assuming we can detect negative zero output of the initial format pass):
  1. deduce the rounding places done by mpd_qformat_spec(), pre-round, change the sign, format again
  2. use the formatted string from the first pass to create a new decimal value (hence already rounded), change the sign, format again

Or, if we can really do (1), then there is no need for two format passes:  if 'z' in spec then pre-round, change the sign if negative zero, format

Reply:
I've corrected rounding of g/e/%

What's remaining is the directed rounding failure.  Hopefully formatting the rounded value instead of the original does it...

Reply:
works! 🤞